vendor/github.com/opencontainers/runc/libcontainer/cgroups/utils.go

```diff
@@ -49,6 +49,24 @@ func IsCgroup2UnifiedMode() bool {
 	return isUnified
 }
 
+// IsCgroup2HybridMode returns whether we are running in cgroup v2 hybrid mode.
+func IsCgroup2HybridMode() bool {
+	isHybridOnce.Do(func() {
+		var st unix.Statfs_t
+		err := unix.Statfs(hybridMountpoint, &st)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// ignore the "not found" error
+				isHybrid = false
+				return
+			}
+			panic(fmt.Sprintf("cannot statfs cgroup root: %s", err))
```

Comment:
this is called from an init() func in vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/fs.go, so anything that links this in will panic if any error that IsNotExist is encountered on startup (like a permissions error)

doesn't that seem problematic?

Reply:
From what I can tell from a git grep, the only components that import this are the Kubelet and a few test files. I think the risk is low, and we could fix this in a patch version backport bump of runc to follow?

<details>

```
ehashman@red-dot:~/src/k8s$ git grep 'github.com/opencontainers/runc/libcontainer/cgroups'
cmd/kubelet/app/server_linux.go:        libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/cgroup_manager_linux.go: libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/cgroup_manager_linux.go: cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
pkg/kubelet/cm/cgroup_manager_linux.go: cgroupfs2 "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
pkg/kubelet/cm/cgroup_manager_linux.go: "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
pkg/kubelet/cm/cgroup_manager_linux.go: cgroupsystemd "github.com/opencontainers/runc/libcontainer/cgroups/systemd"
pkg/kubelet/cm/container_manager_linux.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/container_manager_linux.go:      cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
pkg/kubelet/cm/container_manager_linux.go:      cgroupfs2 "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
pkg/kubelet/cm/container_manager_linux_test.go: "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/helpers_linux.go:        libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/pod_container_manager_linux.go:  libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/qos_container_manager_linux.go:  libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/cm/qos_container_manager_linux.go:  cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
pkg/kubelet/cm/util/cgroups_linux.go:   libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/eviction/threshold_notifier_linux.go:       libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/kuberuntime/kuberuntime_container_linux.go: libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
pkg/kubelet/kuberuntime/kuberuntime_container_linux.go: cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:    cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
pkg/util/oom/oom_linux_test.go: "github.com/opencontainers/runc/libcontainer/cgroups"
test/e2e_node/resource_collector.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
test/e2e_node/utils_linux.go:   libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/container/common/helpers.go:  "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/container/containerd/handler.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/container/crio/handler.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/container/libcontainer/handler.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/container/libcontainer/handler.go:    "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/google/cadvisor/container/libcontainer/helpers.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/container/libcontainer/helpers.go:    fs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
vendor/github.com/google/cadvisor/container/libcontainer/helpers.go:    fs2 "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/google/cadvisor/container/raw/handler.go:     "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/manager/manager.go:   "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/resctrl/utils.go:     "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/google/cadvisor/resctrl/utils.go:     "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/google/cadvisor/utils/sysfs/sysfs.go:// isCPUOnline is copied from github.com/opencontainers/runc/libcontainer/cgroups/fs and modified to suite cAdvisor
vendor/github.com/opencontainers/runc/libcontainer/cgroups/ebpf/devicefilter/devicefilter.go:   devicesemulator "github.com/opencontainers/runc/libcontainer/cgroups/devices"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/blkio.go: "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/cpu.go:   "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/cpu.go:   "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/cpuacct.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/cpuacct.go:       "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/cpuset.go:        "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/cpuset.go:        "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/devices.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/devices.go:       cgroupdevices "github.com/opencontainers/runc/libcontainer/cgroups/devices"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/freezer.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/fs.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/fs.go:    "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/hugetlb.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/hugetlb.go:       "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/memory.go:        "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/memory.go:        "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/name.go:  "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/net_cls.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/net_prio.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/perf_event.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/pids.go:  "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/pids.go:  "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/cpu.go:  "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/cpu.go:  "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/cpuset.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/create.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/devices.go:      "github.com/opencontainers/runc/libcontainer/cgroups/ebpf"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/devices.go:      "github.com/opencontainers/runc/libcontainer/cgroups/ebpf/devicefilter"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/freezer.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/fs2.go:  "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/fs2.go:  "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/hugetlb.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/hugetlb.go:      "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/io.go:   "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/memory.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/memory.go:       "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/pids.go: "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs2/pids.go: "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/fscommon/utils.go:   "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/common.go:   cgroupdevices "github.com/opencontainers/runc/libcontainer/cgroups/devices"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/unsupported.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/v1.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/v1.go:       "github.com/opencontainers/runc/libcontainer/cgroups/fs"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/v2.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/v2.go:       "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/opencontainers/runc/libcontainer/configs/validate/validator.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/container_linux.go:  "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/factory_linux.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/factory_linux.go:    "github.com/opencontainers/runc/libcontainer/cgroups/fs"
vendor/github.com/opencontainers/runc/libcontainer/factory_linux.go:    "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/opencontainers/runc/libcontainer/factory_linux.go:    "github.com/opencontainers/runc/libcontainer/cgroups/systemd"
vendor/github.com/opencontainers/runc/libcontainer/init_linux.go:       "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/notify_linux_v2.go:  "github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
vendor/github.com/opencontainers/runc/libcontainer/process_linux.go:    "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/process_linux.go:    "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/opencontainers/runc/libcontainer/rootfs_linux.go:     "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/github.com/opencontainers/runc/libcontainer/rootfs_linux.go:     "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
vendor/github.com/opencontainers/runc/libcontainer/stats_linux.go:      "github.com/opencontainers/runc/libcontainer/cgroups"
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/devices
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/ebpf
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/ebpf/devicefilter
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/fs
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/fs2
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/fscommon
vendor/modules.txt:github.com/opencontainers/runc/libcontainer/cgroups/systemd
```

</details>

Reply:
`go list -deps` for each cmd directory shows:

```
ehashman@red-dot:~/src/k8s$ for dir in cmd/*; do cd $dir; go list -deps | grep -q 'github.com/opencontainers/runc/libcontainer/cgroups/fs' && echo $dir; cd - &> /dev/null; done
cmd/genkubedocs
cmd/genman
cmd/kubelet
cmd/kubemark
```

with deps... Not sure how kubemark is pulling this in, will that be a problem?

Reply:
kubemark runs a mock kubelet so basically has the same deps. If CI is passing, then I don't think it's a problem, but check with the scaleability team

Reply:
I don't know if anything can function if there's no access to `/sys/fs/cgroup`...

But this is a valid point, and we can do better.